pkg/operator/controller/service-dns/controller.go

```diff
@@ -0,0 +1,230 @@
+package service_dns
+
+import (
+	"context"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	"github.com/openshift/cluster-ingress-operator/pkg/resources/dnsrecord"
+
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"k8s.io/client-go/tools/record"
+
+	corev1 "k8s.io/api/core/v1"
+
+	iov1 "github.com/openshift/api/operatoringress/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const controllerName = "service_dns_controller"
+
+var log = logf.Logger.WithName(controllerName)
+
+// NewUnmanaged creates and returns a controller that watches services that have
+// the "dns.openshift.io/domain" annotation and creates dnsrecord objects for
+// them.  This is an unmanaged controller, which means that the manager does not
+// start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		config:   config,
+		client:   mgr.GetClient(),
+		cache:    mgr.GetCache(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	isServiceNeedingDNS := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		svc := o.(*corev1.Service)
+		return len(svc.Labels["gateway.istio.io/managed"]) != 0
```

Comment:
I haven't looked at that.  Is there a reasonable label to key off of for `istio-ingressgateway`?  I guess the SMCP could explicitly tell OSSM to add the `gateway.istio.io/managed` label.

Reply:
Hmm. Good question. Here's what I have for `istio-ingressgateway` service:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"maistra.io/mesh-generation":"development-1"},"labels":{"app":"istio-ingressgateway","app.kubernetes.io/component":"istio-ingress","app.kubernetes.io/instance":"openshift-ingress","app.kubernetes.io/managed-by":"maistra-istio-operator","app.kubernetes.io/name":"istio-ingress","app.kubernetes.io/part-of":"istio","app.kubernetes.io/version":"development-1","istio":"ingressgateway","istio.io/rev":"openshift-gateway","maistra-version":"2.4.0","maistra.io/owner":"openshift-ingress","maistra.io/owner-name":"openshift-gateway","release":"istio"},"name":"istio-ingressgateway","namespace":"openshift-ingress","ownerReferences":[{"apiVersion":"maistra.io/v2","blockOwnerDeletion":true,"controller":true,"kind":"ServiceMeshControlPlane","name":"openshift-gateway","uid":"74d05a19-ffab-4bef-a25f-e1c0d0e2237f"}]},"spec":{"ports":[{"name":"status-port","port":15021,"protocol":"TCP","targetPort":15021},{"name":"http2","port":80,"protocol":"TCP","targetPort":8080},{"name":"https","port":443,"protocol":"TCP","targetPort":8443}],"selector":{"app":"istio-ingressgateway","istio":"ingressgateway"},"type":"LoadBalancer"}}
    maistra.io/mesh-generation: development-1
  creationTimestamp: "2023-02-22T14:53:27Z"
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
  labels:
    app: istio-ingressgateway
    app.kubernetes.io/component: istio-ingress
    app.kubernetes.io/instance: openshift-ingress
    app.kubernetes.io/managed-by: maistra-istio-operator
    app.kubernetes.io/name: istio-ingress
    app.kubernetes.io/part-of: istio
    app.kubernetes.io/version: development-1
    istio: ingressgateway
    istio.io/rev: openshift-gateway
    maistra-version: 2.4.0
    maistra.io/owner: openshift-ingress
    maistra.io/owner-name: openshift-gateway
    release: istio
  name: istio-ingressgateway
  namespace: openshift-ingress
  ownerReferences:
  - apiVersion: maistra.io/v2
    blockOwnerDeletion: true
    controller: true
    kind: ServiceMeshControlPlane
    name: openshift-gateway
    uid: 74d05a19-ffab-4bef-a25f-e1c0d0e2237f
  resourceVersion: "100493"
  uid: 1d5b3120-ce7d-492a-9603-76be1144aad7
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 172.30.32.183
  clusterIPs:
  - 172.30.32.183
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: status-port
    nodePort: 30200
    port: 15021
    protocol: TCP
    targetPort: 15021
  - name: http2
    nodePort: 31149
    port: 80
    protocol: TCP
    targetPort: 8080
  - name: https
    nodePort: 31413
    port: 443
    protocol: TCP
    targetPort: 8443
  selector:
    app: istio-ingressgateway
    istio: ingressgateway
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: a1d5b3120ce7d492a960376be1144aad-291825524.us-east-1.elb.amazonaws.com
```

> I guess the SMCP could explicitly tell OSSM to add the gateway.istio.io/managed label.

I'm guessing you are referring to `smcp.spec.gateways.ingress.service.metadata.labels`. That seems preferable, since that's a value we can control. It doesn't look like `smcp.spec.gateways.ingress.service.metadata.labels` applies to auto-deployment services just as an FYI.

My personal gut feeling is that we should support the default `istio-ingressgateway` in the same way we support auto-deployments, otherwise it's a bit of a second class citizen. I think this would be pretty trivial right? just add `smcp.spec.gateways.ingress.service.metadata.labels` and it should work?